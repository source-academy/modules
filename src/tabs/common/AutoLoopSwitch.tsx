/* [Imports] */
import { Switch, type SwitchProps } from '@blueprintjs/core';
import React from 'react';

export type AutoLoopSwitchProps = {
  isAutoLooping: boolean,
} & Omit<SwitchProps, 'checked' | 'style'>;

/* [Main] */
export default class AutoLoopSwitch extends React.Component<AutoLoopSwitchProps> {
  render() {
    return <Switch
      style={{
      // Remove default bottom margin
        marginBottom: '0px',

        // Prevent label from wrapping at every letter when width is small
        whiteSpace: 'nowrap',
      }}

      label="Auto Loop"
      checked={ this.props.isAutoLooping }
      {...this.props}
    />;
  }
}
