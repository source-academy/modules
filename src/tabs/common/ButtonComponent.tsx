import { AnchorButton, Button, Intent } from '@blueprintjs/core';
import type { MouseEventHandler, ReactNode } from 'react';

const defaultOptions = {
  className: '',
  fullWidth: false,
  iconOnRight: false,
  intent: Intent.NONE,
  minimal: true
};

type Props = {
  onClick?: MouseEventHandler<HTMLElement>,
  disabled?: boolean,
  children?: ReactNode,
};

const ButtonComponent = (props: Props) => {
  const buttonProps = {
    ...defaultOptions,
    ...props
  };
  return props.disabled
    ? (
      <AnchorButton {...buttonProps} />
    )
    : (
      <Button {...buttonProps} />
    );
};
export default ButtonComponent;
