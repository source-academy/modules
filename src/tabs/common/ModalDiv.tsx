import type { MouseEventHandler, ReactElement } from 'react';

const containerStyle = {
  position: 'fixed',
  zIndex: 999,
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  margin: 'auto',
  padding: '0.5rem',
  display: 'flex',
  flexDirection: 'column',
  alignContent: 'center',
  borderRadius: '5px',
  backgroundColor: 'rgb(199, 196, 196)',
} as React.CSSProperties;

const bodyStyle = {
  backgroundColor: '#fffafa',
  border: '2px solid black',
  borderRadius: '5px',
  boxShadow: '0px 0px 5px inset gray',
  height: '100%',
  overflowY: 'auto',
  padding: '0.75rem',
} as React.CSSProperties;

const backdropStyle = {
  position: 'fixed',
  zIndex: 90,
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  width: '100%',
  height: '100%',
  backgroundColor: 'rgba(0, 0, 0, 0.4)',
  filter: 'blur(10px)',
} as React.CSSProperties;

interface ModalProp {
  open: boolean
  height: string
  width: string
  handleClose: MouseEventHandler
  children: ReactElement
}
const Modal = ({ open, height, width, children, handleClose }: ModalProp) => (
  <>
    {open && (
      <>
        <div style={{
          height,
          width,
          ...containerStyle,
        }}>
          <div style={bodyStyle}>{children}</div>
        </div>
        <div style={backdropStyle} onClick={handleClose} />
      </>
    )}
  </>
);

export default Modal;
