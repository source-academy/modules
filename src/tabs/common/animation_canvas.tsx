import { Button, Icon, Slider, Switch } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Tooltip2 } from '@blueprintjs/popover2';
import React from 'react';
import { type glAnimation } from '../../typings/anim_types';
import WebGLCanvas from './webgl_canvas';

type AnimCanvasProps = {
  animation: glAnimation;
};

type AnimCanvasState = {
  /** Timestamp of the animation */
  animTimestamp: number;

  /** Boolean value indicating if the animation is playing */
  isPlaying: boolean;

  /** Previous value of `isPlaying` */
  wasPlaying: boolean;

  /** Boolean value indicating if auto play is selected */
  autoPlay: boolean;
};

/**
 * Canvas to display glAnimations
 */
// For some reason, I can't get this component to build
// with the blueprint/js components if it's located in
// another file so it's here for now
export default class AnimationCanvas extends React.Component<
AnimCanvasProps,
AnimCanvasState
> {
  private canvas: HTMLCanvasElement | null;

  /**
   * The duration of one frame in milliseconds
   */
  private readonly frameDuration: number;

  /**
   * The duration of the entire animation
   */
  private readonly animationDuration: number;

  /**
   * Last timestamp since the previous `requestAnimationFrame` call
   */
  private callbackTimestamp: number | null;

  constructor(props: AnimCanvasProps | Readonly<AnimCanvasProps>) {
    super(props);

    this.state = {
      animTimestamp: 0,
      isPlaying: false,
      wasPlaying: false,
      autoPlay: true,
    };

    this.canvas = null;
    this.frameDuration = 1000 / props.animation.fps;
    this.animationDuration = Math.round(props.animation.duration * 1000);
    this.callbackTimestamp = null;
  }

  public componentDidMount() {
    this.drawFrame();
  }

  /**
   * Call this to actually draw a frame onto the canvas
   */
  private drawFrame = () => {
    if (this.canvas) {
      const frame = this.props.animation.getFrame(
        this.state.animTimestamp / 1000,
      );
      frame.draw(this.canvas);
    }
  };

  private reqFrame = () => requestAnimationFrame(this.animationCallback);

  private startAnimation = () => this.setState(
    {
      isPlaying: true,
    },
    this.reqFrame,
  );

  private stopAnimation = () => this.setState(
    {
      isPlaying: false,
    },
    () => {
      this.callbackTimestamp = null;
    },
  );

  /**
   * Callback to use with `requestAnimationFrame`
   */
  private animationCallback = (timeInMs: number) => {
    if (!this.canvas || !this.state.isPlaying) return;

    if (!this.callbackTimestamp) {
      this.callbackTimestamp = timeInMs;
      this.drawFrame();
      this.reqFrame();
      return;
    }

    const currentFrame = timeInMs - this.callbackTimestamp;

    if (currentFrame < this.frameDuration) {
      // Not time to draw a new frame yet
      this.reqFrame();
      return;
    }

    this.callbackTimestamp = timeInMs;
    if (this.state.animTimestamp >= this.animationDuration) {
      // Animation has ended
      if (this.state.autoPlay) {
        // If autoplay is active, reset the animation
        this.setState(
          {
            animTimestamp: 0,
          },
          this.reqFrame,
        );
      } else {
        // Otherwise, stop the animation
        this.stopAnimation();
      }
    } else {
      // Animation hasn't ended, so just draw the next frame
      this.setState(
        (prev) => ({
          animTimestamp: prev.animTimestamp + currentFrame,
        }),
        () => {
          this.drawFrame();
          this.reqFrame();
        },
      );
    }
  };

  /**
   * Play button click handler
   */
  private onPlayButtonClick = () => {
    if (this.state.isPlaying) {
      this.stopAnimation();
    } else {
      this.startAnimation();
    }
  };

  /**
   * Reset button click handler
   */
  private onResetButtonClick = () => {
    this.setState(
      {
        animTimestamp: 0,
      },
      () => {
        if (!this.state.isPlaying) this.drawFrame();
      },
    );
  };

  /**
   * Slider value change handler
   * @param newValue New value of the slider
   */
  private onSliderChange = (newValue: number) => {
    this.callbackTimestamp = null;
    this.setState(
      (prev) => ({
        wasPlaying: prev.isPlaying,
        isPlaying: false,
        animTimestamp: newValue,
      }),
      this.drawFrame,
    );
  };

  /**
   * Handler triggered when the slider is clicked off
   */
  private onSliderRelease = () => {
    this.setState(
      (prev) => ({
        isPlaying: prev.wasPlaying,
      }),
      () => {
        if (!this.state.isPlaying) {
          this.callbackTimestamp = null;
        } else {
          this.reqFrame();
        }
      },
    );
  };

  /**
   * Auto play switch handler
   */
  private autoPlaySwitchChanged = () => {
    this.setState((prev) => ({
      autoPlay: !prev.autoPlay,
    }));
  };

  public render() {
    const buttons = (
      <div
        style={{
          marginLeft: '20px',
          marginRight: '20px',
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
        }}
      >
        <div
          style={{
            marginRight: '20px',
          }}
        >
          <Tooltip2 content={this.state.isPlaying ? 'Pause' : 'Play'}>
            <Button onClick={this.onPlayButtonClick}>
              <Icon
                icon={this.state.isPlaying ? IconNames.PAUSE : IconNames.PLAY}
              />
            </Button>
          </Tooltip2>
        </div>
        <Tooltip2 content="Reset">
          <Button onClick={this.onResetButtonClick}>
            <Icon icon={IconNames.RESET} />
          </Button>
        </Tooltip2>
      </div>
    );

    const animSlider = (
      <div
        style={{
          marginTop: '7px',
          flexGrow: 1,
        }}
      >
        <Slider
          value={this.state.animTimestamp}
          onChange={this.onSliderChange}
          onRelease={this.onSliderRelease}
          stepSize={1}
          labelRenderer={false}
          min={0}
          max={this.animationDuration}
        />
      </div>
    );

    return (
      <>
        <div
          style={{
            display: 'flex',
            alignContent: 'center',
            justifyContent: 'center',
          }}
        >
          <WebGLCanvas
            style={{
              flexGrow: 1,
            }}
            ref={(r) => {
              this.canvas = r;
            }}
          />
        </div>
        <div
          style={{
            display: 'flex',
            marginTop: '10px',
            padding: '10px',
            flexDirection: 'row',
            justifyContent: 'stretch',
            alignContent: 'center',
          }}
        >
          {buttons}
          {animSlider}
          <Switch
            style={{
              marginLeft: '20px',
              marginRight: '20px',
              marginTop: '5px',
              whiteSpace: 'nowrap',
            }}
            label="Auto Play"
            onChange={this.autoPlaySwitchChanged}
            checked={this.state.autoPlay}
          />
        </div>
      </>
    );
  }
}
