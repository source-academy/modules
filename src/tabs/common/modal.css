.modal_container {
    position: fixed;
    z-index: 9999;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    margin: auto;
    padding: 0.5rem;
    display: flex;
    flex-direction: column;
    align-content: center;
    border-radius: 5px;
    background-color: rgb(199, 196, 196);
}

.modal_body {
    background-color: #fffafa;
    border: 2px solid black;
    border-radius: 5px;
    box-shadow: 0px 0px 5px inset gray;
    filter: pixelate(50%);
    height: 100%;
    overflow-y: auto;
    padding: 0.75rem;
}

.modal_backdrop {
    position: fixed;
    z-index: 9998;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.4);
}
