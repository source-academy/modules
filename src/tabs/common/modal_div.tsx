import React, { MouseEventHandler, ReactElement } from 'react';

const containerStyle = {
  position: 'fixed',
  zIndex: 999,
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  margin: 'auto',
  padding: '0.5rem',
  display: 'flex',
  flexDirection: 'column',
  alignContent: 'center',
  borderRadius: '5px',
  backgroundColor: 'rgb(199, 196, 196)',
} as React.CSSProperties;

const backdropStyle = {
  position: 'fixed',
  zIndex: 9998,
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  width: '100%',
  height: '100%',
  backgroundColor: 'rgba(0, 0, 0, 0.4)',
} as React.CSSProperties;

interface ModalProp {
  open: boolean
  height: string
  width: string
  handleClose: MouseEventHandler
  children: ReactElement
}

const Modal = ({ open, height, width, children, handleClose }: ModalProp) => (
  <>
    {open && (
      <>
        <div className={'modal_container'} style={{
          height,
          width,
          ...containerStyle,
        }}>
          <div className={'modal_body'}>{children}</div>
        </div>
        <div style={backdropStyle} onClick={handleClose} />
      </>
    )}
  </>
);

export default Modal;
