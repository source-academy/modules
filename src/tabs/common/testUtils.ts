import type { DebuggerContext } from '../../typings/type_helpers';

export const mockDebuggerContext = <T>(state: T, moduleName: string) => ({
  context: {
    moduleContexts: {
      [moduleName]: {
        state,
      },
    },
  },
}) as DebuggerContext;
