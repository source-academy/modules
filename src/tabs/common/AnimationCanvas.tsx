import { Icon, Slider } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Tooltip2 } from '@blueprintjs/popover2';
import React from 'react';
import { type glAnimation } from '../../typings/anim_types';
import AutoLoopSwitch from './AutoLoopSwitch';
import { BP_TAB_BUTTON_MARGIN, BP_TEXT_MARGIN, CANVAS_MAX_WIDTH } from './css_constants';
import PlayButton from './PlayButton';
import WebGLCanvas from './WebglCanvas';
import ButtonComponent from './ButtonComponent';

type AnimCanvasProps = {
  animation: glAnimation;
};

type AnimCanvasState = {
  /** Timestamp of the animation */
  animTimestamp: number;

  /** Boolean value indicating if the animation is playing */
  isPlaying: boolean;

  /** Previous value of `isPlaying` */
  wasPlaying: boolean;

  /** Whether auto loop is enabled */
  isAutoLooping: boolean;

  errored?: any;
};

/**
 * Canvas to display glAnimations.
 *
 * Uses WebGLCanvas internally.
 */
export default class AnimationCanvas extends React.Component<
AnimCanvasProps,
AnimCanvasState
> {
  private canvas: HTMLCanvasElement | null;

  /**
   * The duration of one frame in milliseconds
   */
  private readonly frameDuration: number;

  /**
   * The duration of the entire animation
   */
  private readonly animationDuration: number;

  /**
   * Last timestamp since the previous `requestAnimationFrame` call
   */
  private callbackTimestamp: number | null;

  private reqframeId: number | null;

  constructor(props: AnimCanvasProps | Readonly<AnimCanvasProps>) {
    super(props);

    this.state = {
      animTimestamp: 0,
      isPlaying: false,
      wasPlaying: false,
      isAutoLooping: true,
    };

    this.canvas = null;
    this.frameDuration = 1000 / props.animation.fps;
    this.animationDuration = Math.round(props.animation.duration * 1000);
    this.callbackTimestamp = null;
    this.reqframeId = null;
  }

  public componentDidMount() {
    this.drawFrame();
  }

  /**
   * Call this to actually draw a frame onto the canvas
   */
  private drawFrame = () => {
    if (this.canvas) {
      try {
        const frame = this.props.animation.getFrame(
          this.state.animTimestamp / 1000,
        );
        frame.draw(this.canvas);
      } catch (error) {
        if (this.reqframeId !== null) {
          cancelAnimationFrame(this.reqframeId);
        }

        this.setState({
          isPlaying: false,
          errored: error,
        });
      }
    }
  };

  private reqFrame = () => {
    this.reqframeId = requestAnimationFrame(this.animationCallback);
  };

  private startAnimation = () => this.setState(
    {
      isPlaying: true,
    },
    this.reqFrame,
  );

  private stopAnimation = () => this.setState(
    {
      isPlaying: false,
    },
    () => {
      this.callbackTimestamp = null;
    },
  );

  /**
   * Callback to use with `requestAnimationFrame`
   */
  private animationCallback = (timeInMs: number) => {
    if (!this.canvas || !this.state.isPlaying) return;

    if (!this.callbackTimestamp) {
      this.callbackTimestamp = timeInMs;
      this.drawFrame();
      this.reqFrame();
      return;
    }

    const currentFrame = timeInMs - this.callbackTimestamp;

    if (currentFrame < this.frameDuration) {
      // Not time to draw a new frame yet
      this.reqFrame();
      return;
    }

    this.callbackTimestamp = timeInMs;
    if (this.state.animTimestamp >= this.animationDuration) {
      // Animation has ended
      if (this.state.isAutoLooping) {
        // If auto loop is active, restart the animation
        this.setState(
          {
            animTimestamp: 0,
          },
          this.reqFrame,
        );
      } else {
        // Otherwise, stop the animation
        this.stopAnimation();
      }
    } else {
      // Animation hasn't ended, so just draw the next frame
      this.setState(
        (prev) => ({
          animTimestamp: prev.animTimestamp + currentFrame,
        }),
        () => {
          this.drawFrame();
          this.reqFrame();
        },
      );
    }
  };

  /**
   * Play button click handler
   */
  private onPlayButtonClick = () => {
    if (this.state.isPlaying) this.stopAnimation();
    else this.startAnimation();
  };

  /**
   * Reset button click handler
   */
  private onResetButtonClick = () => {
    this.setState(
      { animTimestamp: 0 },
      () => {
        if (this.state.isPlaying) {
          // Force stop
          this.onPlayButtonClick();
        }

        this.drawFrame();
      },
    );
  };

  /**
   * Slider value change handler
   * @param newValue New value of the slider
   */
  private onSliderChange = (newValue: number) => {
    this.callbackTimestamp = null;
    this.setState(
      (prev) => ({
        wasPlaying: prev.isPlaying,
        isPlaying: false,
        animTimestamp: newValue,
      }),
      this.drawFrame,
    );
  };

  /**
   * Handler triggered when the slider is clicked off
   */
  private onSliderRelease = () => {
    this.setState(
      (prev) => ({
        isPlaying: prev.wasPlaying,
      }),
      () => {
        if (!this.state.isPlaying) {
          this.callbackTimestamp = null;
        } else {
          this.reqFrame();
        }
      },
    );
  };

  /**
   * Auto loop switch onChange callback
   */
  private onSwitchChange = () => {
    this.setState((prev) => ({
      isAutoLooping: !prev.isAutoLooping,
    }));
  };

  public render() {
    return <div
      style={{
        width: '100%',
      }}
    >
      <div
        style={{
          display: 'flex',
          justifyContent: 'center',
        }}
      >
        <div
          style={{
            display: 'flex',
            alignItems: 'center',
            gap: BP_TAB_BUTTON_MARGIN,

            width: '100%',
            maxWidth: CANVAS_MAX_WIDTH,

            paddingTop: BP_TEXT_MARGIN,
            paddingBottom: BP_TEXT_MARGIN,
          }}
        >
          <PlayButton
            isPlaying={this.state.isPlaying}
            disabled={Boolean(this.state.errored)}
            onClick={this.onPlayButtonClick}
          />
          <Tooltip2
            content="Reset"
            placement="top"
          >
            <ButtonComponent
              disabled={Boolean(this.state.errored)}
              onClick={this.onResetButtonClick}
            >
              <Icon icon={ IconNames.RESET } />
            </ButtonComponent>
          </Tooltip2>
          <Slider
            value={ this.state.animTimestamp }
            min={ 0 }
            max={ this.animationDuration }
            stepSize={ 1 }

            labelRenderer={ false }

            disabled={Boolean(this.state.errored)}

            onChange={ this.onSliderChange }
            onRelease={ this.onSliderRelease }
          />
          <AutoLoopSwitch
            isAutoLooping={ this.state.isAutoLooping }
            disabled={Boolean(this.state.errored)}
            onChange={ this.onSwitchChange }
          />
        </div>
      </div>
      <div
        style={{
          display: 'flex',
          justifyContent: 'center',
        }}
      >
        {this.state.errored
          ? (
            <div style={{
              display: 'flex',
              flexDirection: 'column',
              alignItems: 'center',
            }}>
              <div style={{
                display: 'flex',
                flexDirection: 'row',
                alignItems: 'center',
              }}>
                <Icon icon={IconNames.WARNING_SIGN} size={90} />
                <div style={{
                  display: 'flex',
                  flexDirection: 'column',
                  alignItems: 'center',
                  marginBottom: 20,
                }}>
                  <h3>An error occurred while running your animation!</h3>
                  <p style={{ justifySelf: 'flex-end' }}>Here's the details:</p>
                </div>
              </div>
              <code style={{
                color: 'red',
              }}>
                {this.state.errored.toString()}
              </code>
            </div>)
          : (
            <WebGLCanvas
              style={{
                flexGrow: 1,
              }}
              ref={(r) => {
                this.canvas = r;
              }}
            />
          )}
      </div>
    </div>;
  }
}
