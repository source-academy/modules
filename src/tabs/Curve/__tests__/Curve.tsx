import { CurveTab } from ".."
import { animate_3D_curve, animate_curve, draw_3D_connected, draw_connected } from "../../../bundles/curve"
import type { CurveModuleState } from "../../../bundles/curve/types"
import { mockDebuggerContext } from "../../common/testUtils"

test('Curve animations error gracefully', () => {
  const badAnimation = animate_curve(1, 60, draw_connected(200), t => 1 as any)
  const mockContext = mockDebuggerContext<CurveModuleState>({ drawnCurves: [badAnimation] }, 'curve');
  expect(<CurveTab context={mockContext} />)
    .toMatchSnapshot()
})

test('Curve 3D animations error gracefully', () => {
  const badAnimation = animate_3D_curve(1, 60, draw_3D_connected(200), t => 1 as any)
  const mockContext = mockDebuggerContext<CurveModuleState>({ drawnCurves: [badAnimation] }, 'curve');
  expect(<CurveTab context={mockContext} />)
    .toMatchSnapshot()
})