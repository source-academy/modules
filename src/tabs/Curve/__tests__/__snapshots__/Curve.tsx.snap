// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Curve 3D animations error gracefully 1`] = `
<CurveTab
  context={
    {
      "context": {
        "moduleContexts": {
          "curve": {
            "state": {
              "drawnCurves": [
                AnimatedCurve {
                  "angle": 0,
                  "drawer": [Function],
                  "duration": 1,
                  "fps": 60,
                  "func": [Function],
                  "is3D": true,
                  "toReplString": [Function],
                },
              ],
            },
          },
        },
      },
    }
  }
/>
`;

exports[`Curve animations error gracefully 1`] = `
<CurveTab
  context={
    {
      "context": {
        "moduleContexts": {
          "curve": {
            "state": {
              "drawnCurves": [
                AnimatedCurve {
                  "angle": 0,
                  "drawer": [Function],
                  "duration": 1,
                  "fps": 60,
                  "func": [Function],
                  "is3D": false,
                  "toReplString": [Function],
                },
              ],
            },
          },
        },
      },
    }
  }
/>
`;
